### PR TITLE
Add action to force-recalibrate the SCD30

### PR DESCRIPTION
**Pull request by bobobo1618**
## Description:

Adds an automation action to force-recalibrate the SCD30 sensor.

Note: this is my first contribution to this repository so I expect the PR not to be ready as-is. I'd like to leave updating the docs until the PR has seen at least preliminary review.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/scd30/sensor.py [hunk 1]
```diff
@@ -57,3 +61,15 @@ def to_code(config):
...
+    cv.Required(CONF_ID): cv.use_id(SCD30Component),
+    cv.Optional(CONF_FORCED_RECALIBRATION_VALUE): cv.templatable(cv.int_range(min=0, max=0xFFFF, max_included=False)),
+}))
+def set_forced_recalibration_value_to_code(config, action_id, template_arg, args):
+    paren = yield cg.get_variable(config[CONF_ID])
+    var = cg.new_Pvariable(action_id, template_arg, paren)
+    if CONF_FORCED_RECALIBRATION_VALUE in config:
+        template_ = yield cg.templatable(config[CONF_FORCED_RECALIBRATION_VALUE], args, int)
+        cg.add(var.set_forced_recalibration_value(template_))
+    yield var
```
Comment (jesserockz) [inline esphome/components/scd30/sensor.py:75]: Im not sure there is any point in making the value optional as its the whole point of the action...
Other than that, just fix up the lint issues and it should be good.

**Comment by lukasleitsch** [comment]
I have looked for a way to force calibrate the SCD30. Thanks for your PR. It seems there is also a PR to achieve a force calibration. Maybe there is way to merge the two PRs.

https://github.com/esphome/esphome/pull/1237

File: esphome/components/scd30/sensor.py (hunk 1 above)
Comment (bobobo1618) [reply esphome/components/scd30/sensor.py]: Good point, I didn't think too much about that. Done.

**Comment by bobobo1618** [comment]
I made a separate PR for docs: https://github.com/esphome/esphome-docs/pull/866

**Comment by bobobo1618** [comment]
> I have looked for a way to force calibrate the SCD30. Thanks for your PR. It seems there is also a PR to achieve a force calibration. Maybe there is way to merge the two PRs.
> 
> #1237

I had a look at the PR and I don't think there's much to merge. @zeroping can comment if he thinks otherwise but I think the main differences between my PR and his are that I'm adding just the action and the action takes the CO2 concentration as an argument, while his also adds actions for the ASC.

He seemed to think that an argument would be a good idea but didn't know how to do it, so I think that's a good thing to keep.

As for the ASC actions, I don't think they make sense, as the ASC setting is reset to whatever is in the configuration on every boot. Unless there's a way to persist the ASC change across reboots, I think it's more confusing than anything.

**Comment by zeroping** [comment]
Hunh. You know, I'm not actually sure if the ASC setting is kept across reboots. I think I saw this line in the datasheet:

> ASC status is saved in non-volatile memory. When the sensor is powered down while ASC is activated SCD30 will continue with automatic self-calibration after repowering without sending the command.
... and just assumed they meant the value for ASC enable. Maybe they mean the calculated ASC parameters? I'm not sure I've actually tested it.

@bobobo1618, can you confirm if the ASC enable value gets reset to '1' on a power cycle? Mine is a bit tied up, although I might be able to perform a test in the coming weeks.

Point is, if it is stored in persistent memory inside the module, it might be worth keeping asc_disable and asc_enable. Or maybe just setting it every time on power-up, to be sure of the state, although that would be a breaking change for some current SDC30 users.

Either way, I hope one of these two gets merged - but I don't care much which one. @bobobo1618, thanks for poking at this feature with me, even if it's in parallel. I have a system that is doing forced recalibration using outside air and a tiny air pump, and once something is merged, I'd like to add a write-up to the wiki about it, since baseline calibration is so important for CO2 sensors.

**Comment by bobobo1618** [comment]
> I'm not actually sure if the ASC setting is kept across reboots

The _SCD30_ persists it across reboots, _esphome_ does not.

> Or maybe just setting it every time on power-up, to be sure of the state, although that would be a breaking change for some current SDC30 users

That's the current behaviour: https://github.com/esphome/esphome/blob/3afb564a486b7f4744d58639b073575113b7314b/esphome/components/scd30/scd30.cpp#L64

I think leaving the esphome configuration as the source of truth makes sense. That's the easiest logic to reason about: "it does what I say in the config".

**Comment by zeroping** [comment]
> That's the current behaviour:
Yep. I had forgotten that. Then we don't risk breaking anything for anyone!
> I think leaving the esphome configuration as the source of truth makes sense.
Totally agree.

In that case, it sounds like you PR is simpler, but has everything we really need anyways. I guess it would be helpful if I tested your PR against my hardware setup here?

**Comment by bobobo1618** [comment]
By all means, feel free! I have tested it on my own as well though.

File: esphome/components/scd30/automation.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+template<typename... Ts> class SetForcedRecalibrationValueAction : public Action<Ts...> {
+ public:
+  explicit SetForcedRecalibrationValueAction(SCD30Component *scd30) : scd30_(scd30) {}
+
+  TEMPLATABLE_VALUE(uint16_t, forced_recalibration_value);
+
+  void play(Ts... x) override {
+    if (this->forced_recalibration_value_.has_value()) {
+      this->scd30_->set_forced_recalibration_value(this->forced_recalibration_value_.value(x...));
+    }
```
Comment (jesserockz) [inline esphome/components/scd30/automation.h:19]: If it is `required` in config validation, then no need to check if it has a value here
```suggestion
    this->scd30_->set_forced_recalibration_value(this->forced_recalibration_value_.value(x...));
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/scd30/automation.h (hunk 2 above)
Comment (bobobo1618) [reply esphome/components/scd30/automation.h]: Done

File: esphome/components/scd30/scd30.cpp [hunk 3]
```diff
@@ -21,60 +22,65 @@ static const uint16_t SCD30_CMD_TEMPERATURE_OFFSET = 0x5403;
 static const uint16_t SCD30_CMD_SOFT_RESET = 0xD304;
 
 void SCD30Component::setup() {
-  ESP_LOGCONFIG(TAG, "Setting up scd30...");
+  this->set_timeout(1000, [this]() {
```
Comment (jesserockz) [inline esphome/components/scd30/scd30.cpp:25]: Why have you wrapped the whole setup function in a timeout?
If you want it to initialize later on in the cycle you just need to change [the setup priority](https://github.com/esphome/esphome/pull/1402/files#diff-5644f5351ceeef961b3f9ba95cf6165608054f226eba0d7ab6e737d6cdbfd0f3R27)
```

**Comment by geoffrey-vl** [comment]
I may need this feature. How can I help to get it merged?

**Comment by zeroping** [comment]
IIRC, @bobobo1618 and I both made an attempt at this independently, with my attempt over in https://github.com/esphome/esphome/pull/1237. I think we determined that this version looked better than mine.

I still have this hardware running in a window, so I can test. I think we should get this PR merged, as calibration is so important for CO2 sensors.

@bobobo1618, do you have time to fix the branch up? Should I do anything to try to help?

**Comment by bobobo1618** [comment]
I no longer feel motivated to contribute to ESPHome after the way this PR and https://github.com/esphome/esphome/pull/1561 were handled, I'm just going to maintain a small fork for my own use. 

If you want, feel free to clone my branch and create a new PR or take any of my changes into yours.

**Pull request closed without merging**